validation/superchain-version_test.go

```diff
@@ -0,0 +1,135 @@
+package validation
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"golang.org/x/mod/semver"
+)
+
+func checkErr(t *testing.T, err error) {
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+const SOURCE_OF_TRUTH_CHAINID = 10
+const SOURCE_OF_TRUTH_SUPERCHAIN = "mainnet"
+
+// TestContractVersions will check that
+//   - for each declared contract "Foo" in semver.yaml
+//   - for each chain in superchain.OPChain
+//
+// there is a contract address declared for "FooProxy" which has an
+// actual semver matching the desired semver. Actual semvers are
+// read from the L1 chain RPC provider for the chain in question.
+// Desired semvers are read from the contracts on SOURCE_OF_TRUTH_SUPERCHAIN
+// with addresses given by superchain.Addresses[SOURCE_OF_TRUTH_CHAINID]
+func TestContractVersions(t *testing.T) {
+
+	isSemverAcceptable := func(desired, actual string) bool {
+		return semver.Compare(desired, actual) <= 0
+	}
+
+	semverFields := reflect.VisibleFields(reflect.TypeOf(superchain.SuperchainSemver))
+
+	desiredSemver := map[string]string{}
+	for _, field := range semverFields {
+		if field.Name == "SystemConfig" {
+			continue // TODO this is specified in semver.yaml but not in an address list.
+		}
+		proxyContractName := field.Name + "Proxy"
+		r := reflect.ValueOf(superchain.Addresses[SOURCE_OF_TRUTH_CHAINID])
+		proxyContractAddressValue := reflect.Indirect(r).FieldByName(proxyContractName)
+		client, err := ethclient.Dial(superchain.Superchains[SOURCE_OF_TRUTH_SUPERCHAIN].Config.L1.PublicRPC)
+		checkErr(t, err)
+		actualSemver, err := getVersionWithRetries(context.Background(), common.Address(proxyContractAddressValue.Bytes()), client)
+		checkErr(t, err)
+		desiredSemver[field.Name] = actualSemver
+	}
+
+	logOutput, err := json.MarshalIndent(desiredSemver, "", " ")
+	checkErr(t, err)
+
+	t.Logf("Desired Semver from %s/%s : \n %+v", SOURCE_OF_TRUTH_SUPERCHAIN, superchain.OPChains[SOURCE_OF_TRUTH_CHAINID].Name, string(logOutput))
+
+	checkOPChainSatisfiesSemver := func(chain *superchain.ChainConfig) {
+
+		for _, field := range semverFields {
+			if field.Name == "SystemConfig" {
+				continue // TODO this is specified in semver.yaml but not in an address list.
+			}
```

Comment:
This is less than ideal. There's a mismatch in these types which we need to resolve:
https://github.com/ethereum-optimism/superchain-registry/blob/baa49d9c2242562a446254b74c7ebd2d2b5b14ee/superchain/superchain.go#L191-L202

(this is the schema for `semver.yaml`)
and

https://github.com/ethereum-optimism/superchain-registry/blob/baa49d9c2242562a446254b74c7ebd2d2b5b14ee/superchain/superchain.go#L62-L72

(this is what each contributed chain is [asked to provide](https://github.com/ethereum-optimism/superchain-registry/blob/bce4dc12678c705f7c7b0930094d9a21ab0fc674/CONTRIBUTING.md#extras)).

Reply:
The top data type seems to be the canonical one with respect to contract names (which have a version() method, at least). These does seem to imply we want to update `CONTRIBUTING.md` to ask contributors for an address for `SystemConfig`, which it doesn't currently do.